### ISSUE TITLE
Support for the pre-emptive AccessToken refreshment

### ISSUE THREAD

**Issue by sberyozkin** [quarkusio/quarkus#5001]
**Description**
Both access and id tokens can now be supported for the "web-app" applications as part of the code flow. Code for the refresh tokens has also been prototyped. The next step is to use the refresh tokens to pre-emptively refresh ATs when a given access token is close to being expired to avoid the web-app getting 401 when it makes an outbound call with this AT  

**Implementation ideas**
Add a time related property for the adapter to decide when it needs to refresh the token; 
@pedroigor IMHO we'll need to drop the code supporting the injection of RT into the user code, and probably drop `RefreshToken`, because RT can be totally opaque (and as such not map to `JsonWebToken`) but also because it is really an internal adapter task (if we will let the the user code to start refreshing then we'll need to explain how to do it, where to get the IDP address, etc)

IMHO it is not an urgent task

**Comment by sberyozkin** [comment]
@pedroigor OK, but we don't want the users themselves dealing with it, correct ?

**Comment by pedroigor** [comment]
If we provide the necessary capability to automatically refresh tokens, probably yes. But having access to the RT may be useful too.

**Comment by sberyozkin** [comment]
@pedroigor RT is not a token credential really :-). It only enables the refreshment of the AT. So if we do let users see it then we should explain what and how they can do it :-). But that would move us away from the 'make it simple for the users' OIDC strategy. IMHO it will give us more flexibility if we won't start with exposing it, we will always be able to do if there will be a convincing story for it :-)
Thanks

**Comment by sberyozkin** [comment]
@pedroigor Hey Pedro, I thought a bit more about, you are right,  it is not always possible to reliably detect the AT has to be refreshed, there will always be a possibility for the `web-app` doing an outbound call and the AT expiring somewhere in the transit :-).
However I'm still not sure we would ask the users to deal with refreshing the access tokens as there is a bit of complexity involved in the both the request and response validation (esp if OIDC server chooses to return ID token as well).

IMHO, what we should do here is to ship a JAX-RS client filter which will intercept 401/403, check if the AT has expired, and if yes then refresh it and retry the call. This may be tricky to set up (at the level managing the retry HTTP calls with the user code being not aware of it). Or may be we can have an injectable `OidcClient` (which knows how to connect to OIDC, etc) and the user code can then use it to refresh the AT token when needed. Something like that, but I'd consider it be a follow-up issue rather than part of this issue.
As for the `RefreshTokenCredential`, IMHO we'd still be better of drop it and simply have `AccesTokenCredential` have `getRefreshToken` method, because RT is really just a 'helper' for the current AT be refreshed.
Lets discuss it more once we get the scope for working on this issue
Thanks

**Comment by pedroigor** [comment]
I think we should expose it and your suggestion about having `AccesTokenCredential.getRefreshToken` looks very good and much better than having a specific type for the RT.

Like I mentioned before when doing code flow, we should be able to automatically refresh tokens. Where users can enable or disable this capability. At this regard, there are two main points here:

* Users may want to check session state on the server on every single request, so that session invalidation on the server takes immediate effect

* Users may want to always obtain an AT that is valid where the extension will be responsible to make sure the token is always refreshed when the AT is demanded.

For now, I won't work on any in-house client for OAuth2.

**Comment by sberyozkin** [comment]
@pedroigor thanks; sounds good

**Comment by sberyozkin** [comment]
This issue depends on #5288

**Comment by sberyozkin** [comment]
There are 2 RefreshToken classes in the source :-). So another clean up PR is on the way :-). Just copying what I said in the closed #5920.

```
I'm having some thought conflicts about this RefreshToken thing. Like right now we support the injection of it without any support for what they can do, and it is not documented yet, only used in the quickstart to echo it back, so we have a chance to tweak something. As I said, I don;t think we should be asking the users to refresh the tokens directly and then make sure AccessTokenCredentialis synced. Specifically, as discussed at #5002, it is not a standalone credential, and it is confusing to me to see it being injected as such.
What I'd like to do is to update the test to get RT as String from AccessTokenCredential.

By the way, RefreshToken and AccessTokenCredential are in the different packages which is also confusing. I'll move AccessTokenCredential (as well as IdTokenCredential out of runtime one segment up - should be OK since neither of them is doc-ed or used in the quickstarts)
```

**Comment by sberyozkin** [comment]
Hi @pedroigor, I'll just close this one for now as I'm trying to clean a bit - I think we can achieve it anyway while looking at the id token refreshment, and also can be done as part of oidcclient enhancement issue
